diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/ImplicitBackingPropertyRule.kt

```diff
@@ -0,0 +1,112 @@
+ */
+class ImplicitBackingPropertyRule(private val configRules: List<RulesConfig>) : Rule("implicit-backing-property") {
+    private lateinit var emitWarn: ((offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit)
+    private var isFixMode: Boolean = false
+
+    override fun visit(node: ASTNode,
+                       autoCorrect: Boolean,
+                       emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit) {
+        emitWarn = emit
+        isFixMode = autoCorrect
+
+        if (node.elementType == CLASS_BODY) {
+            findAllProperties(node)
+        }
+    }
+
+    @Suppress("UnsafeCallOnNullableType")
+    private fun findAllProperties(node: ASTNode) {
+        val properties = node.getChildren(null).filter { it.elementType == PROPERTY }
+
+        val propsWithBackSymbol = properties
+                .filter { it.getFirstChildWithType(IDENTIFIER)!!.text.startsWith("_") }
+                .map {
+                    it.getFirstChildWithType(IDENTIFIER)!!.text
+                }
+
+        properties.filter { it.hasAnyChildOfTypes(PROPERTY_ACCESSOR) }.forEach {
+            validateAccessors(it, propsWithBackSymbol)
+        }
+    }
+
+    private fun validateAccessors(node: ASTNode, propsWithBackSymbol: List<String>) {
+        val accessors = node.findAllNodesWithSpecificType(PROPERTY_ACCESSOR).filter { it.hasChildOfType(BLOCK) } // exclude get with expression body
+
+        accessors.filter { it.hasChildOfType(GET_KEYWORD) }.forEach { handleGetAccessors(it, node, propsWithBackSymbol) }
+        accessors.filter { it.hasChildOfType(SET_KEYWORD) }.forEach { handleSetAccessors(it, node, propsWithBackSymbol) }
+    }
+
+    @Suppress("UnsafeCallOnNullableType")
+    private fun handleGetAccessors(accessor: ASTNode, node: ASTNode, propsWithBackSymbol: List<String>) {
+        val refExprs = accessor
+                .findAllNodesWithSpecificType(RETURN)
+                .filterNot { it.hasChildOfType(DOT_QUALIFIED_EXPRESSION) }
+                .flatMap { it.findAllNodesWithSpecificType(REFERENCE_EXPRESSION) }
+
+        val localProps = accessor
+                .findAllNodesWithSpecificType(PROPERTY)
+                .map { (it.psi as KtProperty).name!! }
+        // If refExprs is empty then we assume that it returns some constant
+        if (refExprs.isNotEmpty()) {
+            handleReferenceExpressions(node, refExprs, propsWithBackSymbol, localProps)
+        }
+    }
+
+    @Suppress("UnsafeCallOnNullableType")
+    private fun handleSetAccessors(accessor: ASTNode, node: ASTNode, propsWithBackSymbol: List<String>) {
+        val refExprs = accessor.findAllNodesWithSpecificType(REFERENCE_EXPRESSION)
+
+
+        val localProps = accessor
+                .findAllNodesWithSpecificType(PROPERTY)
+                .map { (it.psi as KtProperty).name!! }
+
+        if (refExprs.isNotEmpty()) {
```

Comment:
Does this mean that any reference to non-local property will cause warning? What if some other class property is read for some reason? I think, similar to getters, we should check only left-hand side references in assignments. For example, this code:
```kotlin
val foo
    set(value) {
        if (isDebugMode) log.debug(value)
        field = value
    }
```
shouldn't raise a warning beca use there is no backing property.